tools/cldr-code/src/main/java/org/unicode/cldr/util/PathHeader.java

```diff
@@ -766,6 +751,56 @@ public PathHeader fromPath(final String path, List<String> failures) {
             }
         }
 
+        private PathHeader makePathHeader(RawData data, String path, String alt) {
+            // Caution: each call to PathHeader.Factory.fix changes the value of
+            // PathHeader.Factory.order
+            SectionId newSectionId = SectionId.forString(fix(data.section, 0));
+            String pageIdName = fix(data.page, 0);
+            PageId newPageId;
+            if ("Volume".equals(pageIdName)) {
+                newPageId = getVolumePageId(path);
+            } else {
+                newPageId = PageId.forString(pageIdName);
+            }
+            String newHeader = fix(data.header, data.headerOrder);
+            int newHeaderOrder = (int) order;
+            String codeDashAlt = data.code + (alt == null ? "" : ("-" + alt));
+            String newCode = fix(codeDashAlt, data.codeOrder);
+            long newCodeOrder = order;
+            return new PathHeader(
+                    newSectionId,
+                    newPageId,
+                    newHeader,
+                    newHeaderOrder,
+                    newCode,
+                    newCodeOrder,
+                    suborder,
+                    data.status,
+                    path);
+        }
+
+        private static Set<UnitConverter.UnitSystem> METRIC =
+                Set.of(UnitConverter.UnitSystem.metric, UnitConverter.UnitSystem.metric_adjacent);
+
+        private PageId getVolumePageId(String path) {
+            // Extract the unit from the path. For example, if path is
+            // //ldml/units/unitLength[@type="narrow"]/unit[@type="volume-cubic-kilometer"]/displayName
+            // then extract "volume-cubic-kilometer" which is the long unit id
+            final String longUnitId =
+                    XPathParts.getFrozenInstance(path).findAttributeValue("unit", "type");
+            if (longUnitId == null) {
+                throw new RuntimeException("Missing unit in path " + path);
```

Comment:
We shouldn't throw very generic exceptions, because it makes debugging trickier. If you touch this again, it would be better to have a more specific exception.